tests/test_smart_dispatch.py

```
@@ -95,6 +103,87 @@ def test_main_launch_with_cores_command(self):
         assert_equal(exit_status_100, 2)        
         assert_true(os.path.isdir(self.logs_dir))
 
+    def test_main_launch_with_gpus_command(self):
+        # Actual test
+        exit_status_0 = call(self.launch_command_with_gpus.format(gpus=0), shell=True)
+        exit_status_100 = call(self.launch_command_with_gpus.format(gpus=100), shell=True)
+
+        # Test validation
+        assert_equal(exit_status_0, 0)
+        assert_equal(exit_status_100, 2)
+        assert_true(os.path.isdir(self.logs_dir))
+
+
+    def test_gpu_check(self):
+
+        argv = ['-x', '-g', '2', '-G', '1', '-C', '1', '-q', 'random', '-t', '00:00:10' ,'launch', 'echo', 'testing123']
+
+        # Test if the check fail
+        with self.assertRaises(SystemExit) as context:
+            smartdispatch_script.main(argv=argv)
+
+        self.assertTrue(context.exception.code, 2)
+
+        # Test if the test pass
+        argv[2] = '0'
+
+        try:
+            smartdispatch_script.main(argv=argv)
+        except SystemExit as e:
+            self.fail("The command failed the check, but it was supposed to pass.")
+
+
+    def test_cpu_check(self):
+
+        argv = ['-x', '-c', '2', '-C', '1', '-G', '1', '-t', '00:00:10', '-q', 'random', 'launch', 'echo', 'testing123']
+
+        # Test if the check fail
+        with self.assertRaises(SystemExit) as context:
+            smartdispatch_script.main(argv=argv)
+
+        self.assertTrue(context.exception.code, 2)
+
+        # Test if the test pass
+        argv[2] = '1'
+
+        try:
+            smartdispatch_script.main(argv=argv)
+        except SystemExit as e:
+            self.fail("The command failed the check, but it was supposed to pass.")
```

What happens if you just let it crash on smartdispatch_script.main()? Is it bad because of SystemExit? Do you lose the stacktrace or a valuable error message by doing self.fail?

---

Yes, we lose the the information of the stacktrace when using the self.fail. Otherwise the test crash and we get an error (the rest of the tests runs normally).

I used the self.fail instead of letting the script crash, because I felt it was more inline with what the test was supposed to do (i.e. the test failed because a SystemExit  was thrown, it's not really an unexpected error). But it's right that we lose information that way. Should I simply let it crash instead?

---

I think it is important when a unit test fails that the error message is informative enough that we can easily pin-point the problem and fix it. In this case, I believe both SystemExit and self.fail are just as bad, because SystemExit doesn't write down the stack trace nor informative message about the source of the problem, right? If this is true, then I would do like the other tests where the error is reraised as another exception type. In this case, the reraise procedures should be refactored as a single function to avoid code duplication.